사다리게임 4단계 - 단위 테스트

popsmile opened the pull request:
콘솔 입출력을 다루는 InputView와 ResultView를 제외하고, LadderGame과 Validator의 메소드에 대한 단위테스트를 작성하였습니다.
1. UnitTestLadderGame : 2차원 사다리 배열에 연결된 계단이 있는지, 입력받은 참여자이름이 잘 분리되어 문자열배열로 만들어졌는지 테스트
2. UnitTestValidator : 입력값에 따라 적절한 에러를 던지는지 테스트

popsmile committed:
ResultView 생성자에서 전달인자를 삭제한다

기존에 생성자에 넘겨주던 ladder와 playerNames가 내부 일부의 메소드에서만 사용되기 때문에 printGame()메소드 호출시에만 넘겨주도록 수정한다

popsmile committed:
단위테스트를 위해 UnitTestLadderGame을 생성한다

popsmile committed:
Validator를 테스트하는 메소드를 추가한다

checkValidate()메소드가 입력값에 따라 적절할 에러를 던지는지 테스트한다

popsmile committed:
변수 이름의 단복수를 수정한다

popsmile committed:
LadderGame 메소드 2개의 단위테스트를 추가한다

makeLadder()로 만들어진 2차원 사다리 배열에 연결된 계단이 있는지 검사한다. 입력받은 참여자이름이 getPlayersName()을 통해 잘 분리되어 문자열배열로 만들어졌는지 검사한다.

popsmile committed:
Validator와 LadderGame구조체 각각의 테스트 함수를 분리한다

popsmile committed:
단위테스트 함수명을 수정한다

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -20,7 +20,7 @@ func execute() {
     
     let ladderGame = LadderGame(height: Int(maxHeight)!, names: playerNames)
-    ResultView(ladder:ladderGame.makeLadder(), playerNames:ladderGame.getPlayerNames()).printGame()
+    ResultView().printGame(ladder:ladderGame.makeLadder(), names:ladderGame.getPlayerNames())
```
`ResultView().printGame()` 표현을 보면 억지로 인스턴스를 만들고 있죠.
굳이 인스턴스가 필요없으면 이런 경우는 static 을 사용하는 타입 메서드로 만드는 것도 방법입니다.

popsmile replied:
넵. 인스턴스 메소드와 타입 메소드 차이에 대해 찾아보고 static을 사용하여 수정하였습니다. 궁금한 점은 Validator() 역시 ResultView()처럼 호출하는 부분에서 억지로 인스턴스를 만들고 있는데, ResultView()만 말씀해주신 이유가 따로 있으신가요? 만약 따로 없다면 마찬가지로 수정해볼까 합니다.

godrm replied:
같은 기준에서 다른 곳에 동일하게 적용할 수 있다면 다른 곳에도 적용해보세요 :) 
저도 읽다보면 눈에 보이는 데로 리뷰를 하니까요. 모든걸 다 기억하는건 아닙니다. 본인이 더 잘 알겠죠?

godrm commented on LadderGame/UnitTestLadderGame/UnitTestLadderGame.swift:
```diff
@@ -0,0 +1,49 @@
+        // Put setup code here. This method is called before the invocation of each test method in the class.
+        let maxHeight = "5"
+        let names = "pobi,honux,crong,jk"
+        ladderGame = LadderGame(height: Int(maxHeight)!, names: names)
```
어차피 필요한 값이 5가 들어간 Int 변수면 `maxHeight = 5` 를 하면 되지 않을까요?
코드에서 `!` 가 보이면 개발자가 책임져야 하는 부분입니다. 없으면 없을수록 좋아요.

popsmile committed:
불필요한 전달인자 형변환 과정을 삭제한다

popsmile committed:
ResultView 메소드 모두 static으로 수정한다

main에서 ResultView 인스턴스가 불필요하여 printGame() 및 하위 메소드 모두 타입 메소드로 수정한다

godrm commented on LadderGame/UnitTestLadderGame/UnitTestValidator.swift:
```diff
@@ -0,0 +1,45 @@
+        // height 입력값이 정수형이 아닐 때 -> InputError.notIntType
+        XCTAssertThrowsError(try Validator().checkValidate(height: heightNotAllowed, names: namesAllowed)) {
+            (error) -> Void in XCTAssertEqual(error as? InputError, InputError.notIntType)
+        }
```
테스트를 할 때는 여러개를 한꺼번에 하는게 아니라,
하나의 기능, 하나의 조건을 바꿔가면서 테스트하세요. 
테스트 함수에는 Assert는 확인하려는 기준을 정하는 겁니다. 여러개가 있으면 여러 기준을 만족한다는 거겠죠.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -20,7 +20,7 @@ func execute() {
     
     let ladderGame = LadderGame(height: Int(maxHeight)!, names: playerNames)
-    ResultView().printGame(ladder:ladderGame.makeLadder(), names:ladderGame.getPlayerNames())
+    ResultView.printGame(ladder:ladderGame.makeLadder(), names:ladderGame.getPlayerNames())
```
오늘 피드백 수업에서 했던 것처럼 ladder와 names 에 해당하는 값을 DTO로 만들어서 한꺼번에 전달해보세요.

popsmile committed:
Validator 메소드 모두 static으로 수정한다

Validator 구조체를 호출하는 부분에서 인스턴스가 불필요하여 내부 메소드를 타입메소드로 수정한다

popsmile committed:
UnitTestValidator 메소드를 하나의 테스트단위로 분리한다

하나의 테스트함수에 하나의 Assert 기준만 있도록 수정한다

godrm commented on LadderGame/UnitTestLadderGame/UnitTestLadderGame.swift:
```diff
@@ -0,0 +1,49 @@
+    func testPlayerNamesSepartedProperly() {
+        let playerNames = ladderGame.getPlayerNames()
+        let namesSeparated = ["pobi","honux","crong","jk"]
+        XCTAssertEqual(playerNames, namesSeparated, "Names are NOT separated properly")
```
LadderGame 생성부터 다양한 경우를 확인하기 위해서 LadderGame 생성을 setUp()에 넣지 않고 테스트 함수에 넣는 방법도 있습니다. 
정상적인 경우 말고 `init(height:Int, names:String)` 에 잘못된 입력을 넣는 경우도 다양하게 테스트해보세요.

popsmile replied:
LadderGame 인스턴스 생성 전 매개인자로 넘겨주는 입력값에 대한 예외상황은 모두 Validator에서 걸러지기 때문에, 비정상적인 입력값에 대한 단위테스트는 생략했습니다. 대신, 피드백 주신 내용처럼 잘못된 입력값의 경우를 고려하는 과정에서 새로운 예외상황을 발견하여 Validator에 추가하였습니다 :)

godrm replied:
이 프로그램에서는 InputView > Validator > LadderGame 흐름으로 사용하지만
struct 나 class 로 객체를 만드는 이유는 객체 자체를 재사용하기 위한 겁니다.
모든 경우에 LadderGame을 생성할 때마다 Validator를 거친 값을 넘긴다고 생각하기 어렵습니다. 보이는 init 그대로 사용하게 되는거죠. 그렇게 가정하고 만들지마세요.
만약 그렇게 만든다면 Validator 가 만든 특정한 타입값을 그대로 LadderGame에 넘기도록 만들어야 합니다.

LadderGame 단위 테스트를 한다는 것은 다른 객체와 분맇해서 LadderGame 메소드를 기준으로 테스트하는 겁니다. 따라서 init을 호출했을 때도 안전해야 합니다. 테스트는 결국 설계가 제대로 된것인지 확인하는 과정이라고 생각하고 테스트 코드도 작성하세요.

popsmile replied:
프로그램이 사용하는 흐름에 관계없이 struct 인스턴스를 만들 때에 안전하게 초기화될 수 있도록,  내부 메소드를 수정했습니다. Validator가 만든 특정 타입값을 그대로 LadderGame에 넘기도록 수정하는 것을 먼저 생각했지만, 이 또한 결국 다른 객체와 분리해서 LadderGame 단위 테스트를 시행하는데는 의미를 갖지 못할 것 같아 앞서 말한 방향으로 수정했습니다 :)

popsmile committed:
ResultView에 넘겨줄 LadderGameDTO를 생성한다

popsmile committed:
Unit Test를 통해 새롭게 발견한 NotPositiveType 에러 검증 과정을 추가한다

UnitTestLadderGame에서 LadderGame인스턴스 생성 테스트 메소드를 추가하는 과정에서 height가 양의 정수가 아닐 때 발생하는 에러를 발견한다. 이를 LadderGame 인스턴스 생성 시도 전, 입력값을 받는 단계에서 걸러주기 위하여 Validator 메소드를 수정한다.

popsmile committed:
단위테스트에 testLadderGameDTO메소드를 추가한다

popsmile commented:
1. ResultView에 넘겨줄 값을 LadderGameDTO로 만들어 한꺼번에 전달할 수 있도록 수정했습니다.
2. 사다리 높이에 음의 정수값이 들어오는 예외상황을 처리하는 메소드를 추가했습니다.

popsmile committed:
notPositiveInt 에러를 lessThan2로 수정한다

popsmile committed:
InputError enum을 수정한다

notIntType과 lessThantwo 에러를 notIntMoreThanTwo 에러로 합친다.  참여자 입력값이 없을 경우 던질 noPlayer 에러를 추가한다.

popsmile committed:
LadderGame 구조체 init 메소드를 수정한다

LadderGame 인스턴스 전달인자에 잘못된 값이 전달되어도 정상적 초기화가 가능하도록 Validator를 사용한 내부 검증 메소드를 추가한다.

popsmile committed:
참여자수 검증 메소드명에서 Person을 Player로 수정한다

popsmile committed:
Swift.print문을 기존의 print문으로 수정한다

popsmile committed:
수정된 Validator 메소드를 테스트코드에 반영한다

popsmile committed:
LadderGame구조체 자체로만 테스트 가능하도록 수정한다

Validator와 별개로 LadderGame 구조체를 안전하게 init할 수 있도록 검증 메소드를 추가한다. 이를 단위테스트에 반영한다.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -11,16 +11,18 @@ import Foundation
 
 func execute() {
     let guideMessage = ["참여할 사람의 이름을 입력하세요. (이름은 쉼표(,)로 구분하세요)","최대 사다리 높이는 몇 개인가요?"]
```
이 정도는 나쁘지 않은데, 
struct 나 enum 을 이용한 namespace로 코드화 하는게 더 명확하게 의도를 표현할 수 있을 것 같습니다. 참고하세요.

godrm commented on LadderGame/LadderGame/main.swift:
```diff
@@ -11,16 +11,18 @@ import Foundation
     catch InputError.outOfNameLength { return }
-    catch InputError.notIntType { return }
+    catch InputError.notIntMoreThanTwo { return }
     catch { return }
```
try-catch를 사용하는 경우는 catch 내부에서 꼭 비정상적인 흐름제어를 두는게 좋습니다.
정상적인 경우와 차이가 없으면 try-catch를 잘못 사용하는게 아닌지 의심해볼 필요도 있습니다. 참고하세요.

godrm commented on LadderGame/UnitTestLadderGame/UnitTestLadderGame.swift:
```diff
@@ -0,0 +1,77 @@
+        self.measure {
+            // Put the code you want to measure the time of here.
+        }
+    }
```
안쓰는 함수는 과감하게 지우세요 :)

godrm commented on LadderGame/LadderGame/ResultView.swift:
```diff
@@ -9,32 +9,25 @@
             return
         }
-        Swift.print("     ", terminator: "|")
+        print("     ", terminator: "|")
```
출력할 때 사용하는 하드코딩한 문자열도 의미있는 코드로 바꿔보세요.

godrm commented on LadderGame/LadderGame/LadderGame.swift:
```diff
@@ -40,25 +52,30 @@ struct LadderGame {
+    }
+    
+    func getLadderGameDTO() -> LadderGameDTO {
+        return LadderGameDTO(ladder:makeLadder(), playerNames:getPlayerNames())
```
이 함수는 목표와 결과가 뒤바뀐 것 같습니다.
DTO 객체를 생성하는게 목표가 아니라 makeLadder()를 하는게 목표고, 그래서 만들어진 결과가 DTO 객체인거죠. 
따라서 함수 이름도 getLadderGameDTO 보다는 오히려 makeLadder가 더 의미있는 것 같습니다.
항상 로직의 핵심은 내부 데이터를 처리하는 행동 - 메소드가 되어야 합니다. 함수 이름도 같은 맥락에서 이해해주세요.

godrm commented on LadderGame/LadderGame/Validator.swift:
```diff
@@ -9,28 +9,43 @@
+        guard isIntMoreThanTwo(height:height) else {
+            print("최대 사다리 높이는 2 이상의 정수로 입력해주세요.")
+            throw InputError.notIntMoreThanTwo
         }
```
이 함수는 return 값이 없이 throws 만 하는 함수네요.
그러면 그냥 return 값으로 구분할 수 있지 않을까요?
throws 로 catch 하는것은 (보통 리턴값이 있는 경우) 정상적인 흐름이 아니라 매우 비정상적인 경우에 쓰는게 보통입니다.

- 덧붙여서) 여기처럼 무언가 판단하는 함수에서 출력하는 것은 좋지 않습니다.
판단하는 결과에 따라서 상위에 호출한 로직에서 흐름을 바꾸던지, 에러를 출력하든지, 다른 화면으로 넘어가든지 동작을 처리하도록 하는게 좋습니다. 
그래서 main에서 catch가 아무것도 안하고 있는거죠.

popsmile replied:
넵. 단순히 에러라는 사실에 집중하다 throws, try-catch 구문을 잘못 이해하고 사용했던 것 같습니다. InputError enum의 값 타입을 String으로 변경하고 담긴 내용을 출력하도록 코드를 수정했습니다.

popsmile committed:
ResultView에서 사다리를 출력할 때 사용하는 문자열을 상수로 선언한다

popsmile committed:
guidMessage enum을 생성한다

popsmile committed:
단위테스트에서 사용하지 않는 testPerformanceExample 메소드를 삭제한다

popsmile committed:
InputError를 처리하는 부분에서 try-catch 구문을 삭제한다

InputError 메세지를 담은 enum을 생성하여 에러를 넘기는 방식이 아닌, 리턴받아 처리하도록 수정한다.

popsmile committed:
LadderGameDTO 를 리턴하는 메소드명을 수정한다

메소드의 본 목적은 사다리를 만들기위함이고, 이를 위해 LadderGameDTO를 리턴하는 것이므로 메소드의 목적에 맞게 네이밍을 변경한다.

popsmile committed:
내부에서만 사용되는 메소드를 private으로 수정한다

popsmile committed:
수정된  Validator 메소드에 따라 외부 호출부분 및 단위테스트를 수정한다

popsmile commented:
피드백을 바탕으로 수정한 내용입니다 :)
1. 하드코딩으로 되어있던 부분을 의미있는 상수나 enum으로 선언하였습니다. 
2. 메소드의 본목적을 나타내도록 네이밍을 수정하였습니다. 
3. 잘못된 입력값을 다루는 부분에서 try-catch 구문을 삭제하고, InputError enum을 리턴받아 처리하도록 수정하였습니다.

godrm commented:
수고하셨습니다~
